src/cache_entry.cc

```diff
@@ -0,0 +1,349 @@
+    auto buffer = ToBytes(output);
+    if (!buffer.has_value()) {
+      return Status(
+          Status::Code::INTERNAL, "failed to convert output to bytes");
+    }
+    bool copy = false;  // ToBytes allocated new memory, we pass it through
+    AddBuffer(buffer.value(), copy);
+  }
+
+  return Status::Success;
+}
+
+Status
+CacheEntryItem::ToResponse(InferenceResponse* response)
+{
+  if (!response) {
+    return Status(Status::Code::INTERNAL, "response was nullptr");
+  }
+
+  const auto buffers = Buffers();
+  for (const auto& [base, byte_size] : buffers) {
+    if (!base) {
+      return Status(Status::Code::INTERNAL, "buffer was nullptr");
+    }
+    auto opt_cache_output =
+        FromBytes({static_cast<std::byte*>(base), byte_size});
+    if (!opt_cache_output.has_value()) {
+      return Status(
+          Status::Code::INTERNAL, "failed to convert bytes to response output");
+    }
+    const auto& cache_output = opt_cache_output.value();
+
+    InferenceResponse::Output* response_output = nullptr;
+    RETURN_IF_ERROR(response->AddOutput(
+        cache_output.name_, cache_output.dtype_, cache_output.shape_,
+        &response_output));
+
+    if (!response_output) {
+      return Status(
+          Status::Code::INTERNAL,
+          "InferenceResponse::Output pointer as nullptr");
+    }
+
+    TRITONSERVER_MemoryType memory_type = TRITONSERVER_MEMORY_CPU;
+    int64_t memory_type_id = 0;
+
+    // Allocate buffer for inference response
+    void* output_buffer;
+    RETURN_IF_ERROR(response_output->AllocateDataBuffer(
+        &output_buffer, cache_output.byte_size_, &memory_type,
+        &memory_type_id));
+
+    if (memory_type != TRITONSERVER_MEMORY_CPU &&
+        memory_type != TRITONSERVER_MEMORY_CPU_PINNED) {
+      return Status(
+          Status::Code::INTERNAL,
+          "Only input buffers in CPU memory are allowed in cache currently");
+    }
+
+    if (!output_buffer) {
+      return Status(
+          Status::Code::INTERNAL,
+          "failed to allocate buffer for output '" + cache_output.name_ + "'");
+    }
+    // COPY: cached output buffer to allocated response output buffer
+    memcpy(output_buffer, cache_output.buffer_, cache_output.byte_size_);
+  }
+
+  return Status::Success;
+}
+
+std::optional<Buffer>
+CacheEntryItem::ToBytes(const InferenceResponse::Output& output)
+{
+  // Fetch output buffer details
+  const void* output_base = nullptr;
+  size_t output_byte_size = 0;
+  TRITONSERVER_MemoryType memory_type = TRITONSERVER_MEMORY_CPU;
+  int64_t memory_type_id = 0;
+  void* userp = nullptr;
+  RETURN_NULLOPT_IF_STATUS_ERROR(output.DataBuffer(
+      &output_base, &output_byte_size, &memory_type, &memory_type_id, &userp));
+
+  // DLIS-2673: Add better memory_type support
+  if (memory_type != TRITONSERVER_MEMORY_CPU &&
+      memory_type != TRITONSERVER_MEMORY_CPU_PINNED) {
+    LOG_ERROR
+        << "Only input buffers in CPU memory are allowed in cache currently";
```

Comment:
Does this message gets printed on every inference run? Instead of logging the error, do you think it would be possible to return the error and respond with error for request execution?